File: hot100/Solution0131.cpp
#include <vector>
#include <string>
using namespace std;
// 给你一个字符串 s，请你将 s 分割成一些 子串，使每个子串都是 回文串 。返回 s 所有可能的分割方案。

// 示例 1：

// 输入：s = "aab"
// 输出：[["a","a","b"],["aa","b"]]
// 示例 2：

// 输入：s = "a"
// 输出：[["a"]]

// 提示：

// 1 <= s.length <= 16
// s 仅由小写英文字母组成
// 方法一：暴力回溯
class Solution
{
public:
    vector<vector<string>> partition(string s)
    {
        vector<vector<string>> result;
        vector<string> path;
        backtrack(s, 0, path, result);
        return result;
    }

private:
    void backtrack(string &s, int index, vector<string> path, vector<vector<string>> &result)
    {
        if (index == s.size())
        {
            if (isPalindrome(path))
                result.push_back(path);
            return;
        }
        for (int i = index; i < s.size(); i++)
        {
            path.push_back(s.substr(index, i - index + 1));
            backtrack(s, i + 1, path, result);
            path.pop_back();
        }
    }
    bool isPalindrome(vector<string> &path)
    {
        for (const string &s : path)
        {
            int left = 0, right = s.size() - 1;
            while (left < right)
            {
                if (s[left] != s[right])
                    return false;
                left++;
                right--;
            }
        }
        return true;
    }
};

// 方法二：回溯+动态规划.
// 当前搜索到字符串的第 i 个字符，且 s[0..i−1] 位置的所有字符已经被分割成若干个回文串
// 遍历j，使得s[i..j] 是一个回文串
// i>j，空字符串：True
// i=j，单字符串：True
// i<j，s[i] = s[j] && [i+1,j-1]是回文。
// 因此，使用cache[i][j]预处理s[i,j]是不是回文。之后回溯的时候就不需要再判断是否是回文了
class Solution2
{
private:
    vector<vector<bool>> cache;

public:
    vector<vector<string>> partition(string s)
    {
        int n = s.size();
        cache.assign(n, vector<bool>(n, true));
        vector<vector<string>> result;
        vector<string> path;
        for (int i = n - 1; i >= 0; i--) // 从字符串末尾开始
            for (int j = i + 1; j < n; j++)
                cache[i][j] = (s[i] == s[j]) && cache[i + 1][j - 1];

        backtrack(s, 0, path, result);
        return result;
    }

private:
    void backtrack(const string &s, int index, vector<string> &path, vector<vector<string>> &result)
    {
        if (index == s.size())
        {
            result.push_back(path);
            return;
        }

        for (int i = index; i < s.size(); i++)
        {
            if (cache[index][i])
            {
                path.push_back(s.substr(index, i - index + 1));
                backtrack(s, i + 1, path, result);
                path.pop_back();
            }
        }
    }
};

// 方法三：回溯+记忆化搜索。
// 即不要预处理生成cache，而是中途生成
class Solution3
{
private:
    vector<vector<int>> cache; // 0表示还未处理，1表示是回文，-1表示不是回文
public:
    vector<vector<string>> partition(string s)
    {
        int n = s.size();
        vector<vector<string>> result;
        vector<string> path;

        backtrack(s, 0, path, result);
        return result;
    }

private:
    void backtrack(const string &s, int index, vector<string> &path, vector<vector<string>> &result)
    {
        if (index == s.size())
        {
            result.push_back(path);
            return;
        }

        for (int i = index; i < s.size(); i++)
        {
            if (isPalindrome(s, index, i))
            {
                path.push_back(s.substr(index, i - index + 1));
                backtrack(s, i + 1, path, result);
                path.pop_back();
            }
        }
    }
    int isPalindrome(const string &s, int left, int right)
    {
        if (cache[left][right])
            return cache[left][right];
        if (left >= right)
        {
            return cache[left][right] = 1;
        }
        return cache[left][right] = (s[left] == s[right]) ? isPalindrome(s, left + 1, right - 1) : -1;
    }
};

File: hot100/Solution0131.java
package hot100;

import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;

// 给你一个字符串 s，请你将 s 分割成一些 子串，使每个子串都是 回文串 。返回 s 所有可能的分割方案。

public class Solution0131 {
    public List<List<String>> partition(String s) {
        List<List<String>> ans = new ArrayList<>();
        backtrack(ans, new ArrayList<>(), s, 0);
        return ans;
    }

    private void backtrack(List<List<String>> ans, List<String> path, String s, int index) {
        if (index == s.length()) {
            ans.add(new ArrayList<>(path));
            return;
        }
        for (int i = index; i < s.length(); i++) {
            if (isPalindrome(s, index, i)) {
                path.add(s.substring(index, i + 1));
                backtrack(ans, path, s, i + 1);
                path.remove(path.size() - 1);
            }
        }
    }

    private boolean isPalindrome(String s, int left, int right) {
        while (left < right) {
            if (s.charAt(left) != s.charAt(right)) {
                return false;
            }
            left++;
            right--;
        }
        return true;
    }

    private boolean[][] cache2;

    public List<List<String>> partition2(String s) {
        List<List<String>> ans = new ArrayList<>();
        cache2 = new boolean[s.length()][s.length()];
        for (int i = 0; i < s.length(); i++) {
            Arrays.fill(cache2[i], true);
        }
        for (int i = s.length() - 1; i >= 0; i--) {
            for (int j = i + 1; j < s.length(); j++) {
                cache2[i][j] = s.charAt(i) == s.charAt(j) && cache2[i + 1][j - 1];
            }
        }
        backtrack2(ans, new ArrayList<>(), s, 0);
        return ans;
    }

    private void backtrack2(List<List<String>> ans, List<String> path, String s, int index) {
        if (index == s.length()) {
            ans.add(new ArrayList<>(path));
            return;
        }
        for (int i = index; i < s.length(); i++) {
            if (cache2[index][i]) {
                path.add(s.substring(index, i + 1));
                backtrack2(ans, path, s, i + 1);
                path.remove(path.size() - 1);
            }
        }
    }

    private int[][] cache3; // 0: unknown, 1: palindrome, -1: not palindrome

    public List<List<String>> partition3(String s) {
        cache3 = new int[s.length()][s.length()];
        List<List<String>> ans = new ArrayList<>();
        backtrack3(ans, new ArrayList<>(), s, 0);
        return ans;
    }

    private void backtrack3(List<List<String>> ans, List<String> path, String s, int index) {
        if (index == s.length()) {
            ans.add(new ArrayList<>(path));
            return;
        }
        for (int i = index; i < s.length(); i++) {
            if (isPalindrome3(s, index, i)) {
                path.add(s.substring(index, i + 1));
                backtrack3(ans, path, s, i + 1);
                path.remove(path.size() - 1);
            }
        }
    }

    private boolean isPalindrome3(String s, int left, int right) {
        if (cache3[left][right] == 1) {
            return true;
        }
        if (left >= right) {
            cache3[left][right] = 1;
            return true;
        }
        if (s.charAt(left) == s.charAt(right) && isPalindrome3(s, left + 1, right - 1)) {
            cache3[left][right] = 1;
            return true;
        }
        cache3[left][right] = -1;
        return false;
    }

    public static void main(String[] args) {
        Solution0131 solution = new Solution0131();
        String s = "aab";
        List<List<String>> ans = solution.partition(s);
        System.out.println(ans);

        List<List<String>> ans2 = solution.partition2(s);
        System.out.println(ans2);

        List<List<String>> ans3 = solution.partition3(s);
        System.out.println(ans3);
    }
}
